File: 2021/october/day-7/solution1.cpp
class Solution {
    public:
        bool exist(vector<vector<char>>& board, string word) {
             int len = board.size(), ans =0;
            for(int i = 0; i<len;i++){
                for(int j=0; j<board[0].size();j++){
                    // if((board[i][j] == word.at(0)) && solve(board, i, j, 0, word))
                    //     return true;
                    ans += solve(i, j, word, board, 0);
                    if(ans > 0) return true;
                }
                
            }
            
        return false;
    
       }
        
        // bool solve (vector<vector<char>>& board, int i, int j, int count, string &word)
         int solve (int i, int j, string word, vector<vector<char>>& board, int count)
        {
           
             int found =0;
             if(i>=0 and j>=0 and i< board.size() and j < board[0].size() and word[count] == board[i][j] )
             {
                 char temp = word[count];
                 board[i][j] ='*';
                 
                 count++;
                 if(count == word.size()) found +=1;
             
                 else{
                     found += solve(i+1, j, word, board, count);
                     found += solve(i-1, j, word, board, count);
                     found += solve(i, j+1, word, board, count);
                     found += solve(i, j-1, word, board, count);
    
                 }
                 board[i][j] = temp;
                 
             }
             return found;
        
    }
    };

File: 2021/october/day-9/solution1.py
class Node:
        def __init__(self,c,end=False):
            self.child = {}
            self.c = c
            self.end = end
            
    class Trie:
        def __init__(self):
            self.root = Node("/")
            
        def insert(self,word):
            cur = self.root
            for c in word:
                if c not in cur.child: cur.child[c] = Node(c)
                cur = cur.child[c]
            cur.end = True
                    
    
    
    class Solution:
        def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
            lookup = Trie()
            for w in words:
                lookup.insert(w)
                
            def dfs(r,c,cur,word):
                if not (0<=r<n and 0<=c<m): return
                w = board[r][c]
                if w not in cur.child: return
                if cur.child[w].end:
                    ans.add(word+w)
                board[r][c] = " "
                dfs(r+1,c,cur.child[w],word+w)
                dfs(r-1,c,cur.child[w],word+w)
                dfs(r,c+1,cur.child[w],word+w)
                dfs(r,c-1,cur.child[w],word+w)
    
                board[r][c] = w
                
            n = len(board)
            m = len(board[0])
            ans = set()
            
            for r in range(n):
                for c in range(m):
                    dfs(r,c,lookup.root,"")
            
            return ans